### DISCUSSION TITLE
Trouble migrating from CRA 4.0.3 to CRA 5.0.0

### DISCUSSION THREAD

**Discussion by NasirKhalidQ** [tailwindlabs/tailwindcss#7007] [Help]
**What version of Tailwind CSS are you using?**

v3.0.13

**What build tool (or framework if it abstracts the build tool) are you using?**

React.js v17.0.2 with PostCSS and autoprefixer

**What version of Node.js are you using?**

v16.13.0

**What browser are you using?**

Chrome

**What operating system are you using?**

MacOs Monterey v12.1

-----------------------------------------------------
Hello everyone!
First of all a big thank you to the creators! TailwindCSS is the best thing to happen to front-end web dev and I am truly grateful that we have this, I would not have been able to learn CSS so quickly if it wasn't for Tailwind! So thank you for all your efforts!
Secondly, I am having quite a bit of issues when trying to upgrade my company project to v3. I upgraded my react-scripts package from 4.0.3 to 5.0.0 and it takes a long time to compile then just throws an error. 

> Starting the development server...
> 
> 
> <--- Last few GCs --->
> 
> [3817:0x7f8d69800000]    42798 ms: Scavenge 2022.5 (2075.5) -> 2014.6 (2075.4) MB, 76.0 / 0.0 ms  (average mu = 0.249, current mu = 0.207) allocation failure 
> [3817:0x7f8d69800000]    42809 ms: Scavenge 2023.8 (2076.9) -> 2016.9 (2076.7) MB, 3.0 / 0.0 ms  (average mu = 0.249, current mu = 0.207) allocation failure 
> [3817:0x7f8d69800000]    43095 ms: Scavenge 2026.1 (2079.1) -> 2019.7 (2079.7) MB, 279.2 / 0.0 ms  (average mu = 0.249, current mu = 0.207) allocation failure 
> 
> 
> <--- JS stacktrace --->
> 
> FATAL ERROR: Ineffective mark-compacts near heap limit Allocation failed - JavaScript heap out of memory
>  1: 0x107613515 node::Abort() (.cold.1) [/usr/local/bin/node]
>  2: 0x106314989 node::Abort() [/usr/local/bin/node]
>  3: 0x106314aff node::OnFatalError(char const*, char const*) [/usr/local/bin/node]
>  4: 0x1064942c7 v8::Utils::ReportOOMFailure(v8::internal::Isolate*, char const*, bool) [/usr/local/bin/node]
>  5: 0x106494263 v8::internal::V8::FatalProcessOutOfMemory(v8::internal::Isolate*, char const*, bool) [/usr/local/bin/node]
>  6: 0x106635975 v8::internal::Heap::FatalProcessOutOfMemory(char const*) [/usr/local/bin/node]
>  7: 0x1066399bd v8::internal::Heap::RecomputeLimits(v8::internal::GarbageCollector) [/usr/local/bin/node]
>  8: 0x10663629d v8::internal::Heap::PerformGarbageCollection(v8::internal::GarbageCollector, v8::GCCallbackFlags) [/usr/local/bin/node]
>  9: 0x1066337bd v8::internal::Heap::CollectGarbage(v8::internal::AllocationSpace, v8::internal::GarbageCollectionReason, v8::GCCallbackFlags) [/usr/local/bin/node]
> 10: 0x1066324d8 v8::internal::Heap::HandleGCRequest() [/usr/local/bin/node]
> 11: 0x1065decd1 v8::internal::StackGuard::HandleInterrupts() [/usr/local/bin/node]
> 12: 0x1069c2ca8 v8::internal::Runtime_StackGuard(int, unsigned long*, v8::internal::Isolate*) [/usr/local/bin/node]
> 13: 0x106d6cdd9 Builtins_CEntry_Return1_DontSaveFPRegs_ArgvOnStack_NoBuiltinExit [/usr/local/bin/node]
> 14: 0x112d76325 

I am pretty sure this has to do with CRA as even before adding v3 for tailwind it gives the same issue. 
I have upgraded my PostCSS and Autoprefixer to the latest version and have tried the suggestions in other discussions such as removing the package-lock.json and node modules but no avail :(
Any help would be appreciated. I have removed CRACO and changed my package.json to use react-scripts.

Here is my **tailwind.config.js**:
const colors = require("tailwindcss/colors");

> module.exports = {
>   separator: "_",
>   corePlugins: {
>     preflight: false,
>   },
>   content: ["./src/**/*.{js,jsx,ts,tsx}", "./public/index.html"],
>   theme: {
>     fontFamily: {
>       Poppins: ["Poppins", "sans-serif"],
>       Roboto: ["Roboto"],
>     },
>     colors: {
>       current: "currentColor",
>     },
> 
>     extend: {
>       colors: {
>         "outwiths-comfortable-green": "#53C5BD",
>         "outwiths-orange": "#F05322",
>         "outwiths-bg-user": "#FFD6CB",
>         "outwiths-bg-employer": "#C5E3E3",
>         "outwiths-green": "#2BAAAC",
>         "outwiths-007-green": "#007777",
>         "bg-linkedin": "#0A66C2",
>         "outwiths-light-salmon": "#FFAF9F",
>         "outwiths-dark-orange": "#EC2630",
>         "outwiths-hot-pink": "#FF3C5F",
>         "outwiths-black-0": "rgba(0, 0, 0, 0)",
>         "outwiths-black-60": "rgba(0, 0, 0, 0.6)",
>         "share-facebook": "#395185",
>         "share-instagram": "#EB4A65",
>         "share-twitter": "#55ACEE",
>         "share-pinterest": "#CB1F27",
>         "share-linkedin": "#0A66C2",
>         green: colors.emerald,
>         yellow: colors.amber,
>         purple: colors.violet,
>       },
>       fontSize: {
>         xxs: ".5rem",
>         px10: "0.625rem",
>       },
>     },
>   },
> 
>   plugins: [
>     require("@tailwindcss/forms"),
>     require("@tailwindcss/aspect-ratio"),
>   ],
> };

Here is my **package.json**:

> {
>   "name": "outwiths-ui",
>   "version": "0.1.0",
>   "private": true,
>   "dependencies": {
>     "@dnd-kit/core": "^4.0.2",
>     "@dnd-kit/sortable": "^5.1.0",
>     "@headlessui/react": "^1.4.2",
>     "@testing-library/jest-dom": "^5.11.4",
>     "@testing-library/react": "^11.1.0",
>     "@testing-library/user-event": "^12.1.10",
>     "history": "^5.1.0",
>     "react": "^17.0.2",
>     "react-dom": "^17.0.2",
>     "react-pdf": "^5.6.0",
>     "react-router-dom": "^6.0.2",
>     "react-scripts": "^5.0.0",
>     "uuid": "^8.3.2",
>     "web-vitals": "^1.0.1"
>   },
>   "scripts": {
>     "start": "react-scripts start",
>     "build": "react-scripts build",
>     "test": "react-scripts test",
>     "eject": "react-scripts eject"
>   },
>   "eslintConfig": {
>     "extends": [
>       "react-app",
>       "react-app/jest"
>     ]
>   },
>   "browserslist": {
>     "production": [
>       ">0.2%",
>       "not dead",
>       "not op_mini all"
>     ],
>     "development": [
>       "last 1 chrome version",
>       "last 1 firefox version",
>       "last 1 safari version"
>     ]
>   },
>   "devDependencies": {
>     "@tailwindcss/aspect-ratio": "^0.4.0",
>     "@tailwindcss/forms": "^0.4.0",
>     "autoprefixer": "^10.4.2",
>     "postcss": "^8.4.5",
>     "postcss-loader": "^6.2.1",
>     "prettier": "^2.4.1",
>     "tailwindcss": "^3.0.12"
>   }
> }

Here is my **postcss.config.js**:

> module.exports = {
>   plugins: {
>     tailwindcss: {},
>     autoprefixer: {},
>   },
> };

Here is my **index.css**:

> /* ./src/index.css */
> @tailwind base;
> @tailwind components;
> 
> @tailwind utilities;
> @import url("https://fonts.googleapis.com/css2?family=Poppins:wght@100;200;300;400;500;600;700;800;900&display=swap");
> @layer base {
>   h1 {
>     @apply text-3xl;
>   }
>   h2 {
>     @apply text-2xl;
>   }
>   h3 {
>     @apply text-xl;
>   }
>   h4 {
>     @apply text-lg;
>   }
>   h5 {
>     @apply text-base;
>   }
> 
>   p {
>     @apply text-sm;
>   }
> }
> 
> @layer components {
>   .input-styled {
>     @apply mt-2 rounded-lg py-4 bg-gray-100 border-gray-100 shadow-sm focus:border-gray-100 focus:ring-0 focus:ring-gray-100 focus:ring-opacity-50;
>   }
>   .profile-link {
>     @apply bg-gray-100 border-gray-100 focus:border-gray-100 focus:ring-0 focus:ring-gray-100 focus:ring-opacity-50;
>   }
>   .checkbox-orange {
>     @apply text-outwiths-orange w-5 h-5 focus:ring-outwiths-orange self-center focus:ring-opacity-0 focus:ring-0 focus:ring-offset-0 rounded-md;
>   }
> }
> 
> .input-styled p {
>   color: rgb(70, 70, 70);
>   font-size: 1em;
> }
> 
> .divider p {
>   width: 100%;
>   text-align: center;
>   border-bottom: 1px solid;
>   line-height: 0.1em;
>   margin: 10px 0 20px;
> }
> .divider p span {
>   padding: 0 10px;
>   background: #fff;
> }
> .projectContainer {
>   width: 249px;
>   height: 178.06px;
> }
> .linkModal {
>   width: 791px;
>   height: 604px;
> }
> .uploadLink {
>   width: 390px;
>   height: 219px;
> }
> .inputLink {
>   width: 390px;
> }
> .userImage {
>   width: 44px;
>   height: 45px;
> }
> .mainImage {
>   width: 864.84px;
>   height: 618.08px;
> }
> 
> .viewProjectContainer {
>   height: 589.25px;
> }
> 
> .sideImage {
>   width: 143.45px;
>   height: 102.54px;
> }

When i use CRA 5 with a new project and install tailwind in it, it works perfect but I can not upgrade my existing project :( I would not be able to put everything in new one since the project has grown to be quite big now.
I am on a Mac with VS Code and my Node is v16.13.0, my npm version is 8.1.0
Any help from the community is appreciated!

PS: If anyone could create a noob-friendly guide for migrating an existing tailwind v2 project with CRA 4 to tailwind v3 with CRA 5 that would be great!
![image](https://user-images.githubusercontent.com/75773436/148882200-b667079d-bcc5-46d3-8d09-3d446ad2408d.png)

### UPDATE
I have given up on CRA v5 for now and I think only for new projects I will be able to use CRA v5
I have used a solution by Adam in this repo:
https://github.com/adamwathan/cra-tailwind-3
If anyone else is having issues with react-scripts 5.0.0 then just use the existing 4.0.3 which you have and instead of poscss use tailwind cli with the configuration in above repo for tailwind.config.js and package.json. 
You would need to install "concurrently" to make it work with tailwind cli.
Build times are faster and now you can use tailwind v3.

**Comment by webstar-code** [comment]
First make sure to checkout to different branch or push your code to github before migrating, just for safety.
1. First run `npm uninstall @craco/craco  autoprefixer postcss tailwindcss`
    Delete the **craco.config.js file**
    Delete the **tailwind.config.js**

Now just follow the [offical tailwindcss](https://tailwindcss.com/docs/guides/create-react-app) docs
2. Run `npm install -D tailwindcss postcss autoprefixer`
3. `npx tailwindcss init -p`
    copy paste the tailwind.config.js file from tailwindcss docs
4. `npm install react-scripts@latest`
5. Now `npm start`
later paste your previous tailwind.config theme in the **new tailwind.config.js** file

The order of steps may or may not matter but following this steps worked for me

**Reply by NasirKhalidQ** [reply]
Hello friend!
Thank you for suggesting these steps. I have followed them but unfortunately it did not solve my problem.
Since the project is too big with a few dependencies and a lot of customised code I am assuming it is some issue at my end, so none of the suggested fixes are solving the problem :(
I have given up on CRA v5 for now and I think only for new projects I will be able to use CRA v5
I have used a solution by Adam in this repo:
https://github.com/adamwathan/cra-tailwind-3
If anyone else is having issues with react-scripts 5.0.0 then just use the existing 4.0.3 which you have and instead of poscss use tailwind cli with the configuration in above repo for tailwind.config.js and package.json. 
You would need to install "concurrently" to make it work with tailwind cli.
Build times are faster and now you can use tailwind v3.
Thank you again for suggesting the solution! 
Cheers 👍

**Comment by tonnoz** [comment]
your solution works! Thanks!

offtopic: which vscode theme do you use @NasirKhalidQ  👁👄👁?
